rmb-sdk-go/peer/connection_logger.go

```diff
@@ -0,0 +1,88 @@
+package peer
+
+import (
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+// NewLogObserver builds a zerolog-based observer with the approved sampling.
+func NewLogObserver(url string) ConnObserver {
+	// reader/output backpressure: 5s after first
+	readerSampler := zerolog.LevelSampler{WarnSampler: &zerolog.BurstSampler{Burst: 1, Period: 5 * time.Second}}
+	bpReaderLog := log.With().Str("url", url).Logger().Sample(&readerSampler)
+
+	outputSampler := zerolog.LevelSampler{WarnSampler: &zerolog.BurstSampler{Burst: 1, Period: 5 * time.Second}}
+	bpOutLog := log.With().Str("url", url).Logger().Sample(&outputSampler)
+
+	// write spike: 5s after first
+	spikeSampler := zerolog.LevelSampler{WarnSampler: &zerolog.BurstSampler{Burst: 1, Period: 5 * time.Second}}
+	spikeLog := log.With().Str("url", url).Logger().Sample(&spikeSampler)
+
+	return &logObserver{
```

Comment:
I feel they are repeated?

Reply:
If you reuse a single sampler instance, a burst of one category could suppress logs from the others. so using separate zerolog.LevelSampler instances ensures each category has its own rate-limiting bucket.